src/state/slices/stakingDataSlice/stakingDataSlice.test.ts

```diff
@@ -0,0 +1,522 @@
+        const selected = selectActiveStakingOpportunityDataByAssetId(
+          store.getState(),
+          cosmosAccountSpecifier,
+          SHAPESHIFT_VALIDATOR_ADDRESS,
+          cosmosAssetId
+        )
+        expect(selected).toEqual([])
+      })
+
+      it('returns empty array when staking data is loaded but validators are not', async () => {
+        store.dispatch(
+          stakingData.actions.upsertStakingData({
+            accountSpecifier: cosmosAccountSpecifier,
+            stakingData: mockStakingData
+          })
+        )
+
+        const selected = selectActiveStakingOpportunityDataByAssetId(
+          store.getState(),
+          cosmosAccountSpecifier,
+          SHAPESHIFT_VALIDATOR_ADDRESS,
+          cosmosAssetId
+        )
+        expect(selected).toEqual([])
+      })
+
+      describe('validators and staking data are loaded', () => {
+        beforeEach(() => {
+          store.dispatch(
+            stakingData.actions.upsertValidatorData({
+              validators: mockValidatorData
+            })
+          )
+        })
+
+        it('returns correct array', async () => {
+          store.dispatch(
+            stakingData.actions.upsertStakingData({
+              accountSpecifier: cosmosAccountSpecifier,
+              stakingData: mockStakingData
+            })
+          )
+
+          const selected = selectActiveStakingOpportunityDataByAssetId(
+            store.getState(),
+            cosmosAccountSpecifier,
+            SHAPESHIFT_VALIDATOR_ADDRESS,
+            cosmosAssetId
+          )
+          expect(selected).toMatchSnapshot()
+        })
+
+        it('returns correct array with only undelegations', async () => {
+          store.dispatch(
+            stakingData.actions.upsertStakingData({
+              accountSpecifier: cosmosAccountSpecifier,
+              stakingData: mockStakingDataWithOnlyUndelegations
+            })
+          )
+
+          const selected = selectActiveStakingOpportunityDataByAssetId(
+            store.getState(),
+            cosmosAccountSpecifier,
+            SHAPESHIFT_VALIDATOR_ADDRESS,
+            cosmosAssetId
+          )
+          expect(selected).toMatchSnapshot()
+        })
+
+        it('returns correct array with only rewards', async () => {
+          store.dispatch(
+            stakingData.actions.upsertStakingData({
+              accountSpecifier: cosmosAccountSpecifier,
+              stakingData: mockStakingDataWithOnlyRewards
+            })
+          )
+
+          const selected = selectActiveStakingOpportunityDataByAssetId(
+            store.getState(),
+            cosmosAccountSpecifier,
+            SHAPESHIFT_VALIDATOR_ADDRESS,
+            cosmosAssetId
+          )
+          expect(selected).toMatchSnapshot()
+        })
+      })
+    })
+
+    describe('selectNonloadedValidators', () => {
+      it('returns empty array when state is empty', async () => {
```

Comment:
```suggestion
      it('returns empty array on initial state', async () => {
```
I'd go with `initial state` terminology instead, empty is a high-level human concept, but from a structure perspective, the initial state is not an empty object